temporal-sdk/src/main/java/io/temporal/internal/statemachines/WorkflowStateMachines.java

```diff
@@ -705,6 +726,34 @@ private void handleNonStatefulEvent(HistoryEvent event, boolean hasNextEvent) {
       case EVENT_TYPE_WORKFLOW_EXECUTION_CANCEL_REQUESTED:
         callbacks.cancel(event);
         break;
+      case EVENT_TYPE_WORKFLOW_EXECUTION_UPDATE_ADMITTED:
+        WorkflowExecutionUpdateAdmittedEventAttributes admittedEvent =
+            event.getWorkflowExecutionUpdateAdmittedEventAttributes();
+        Message msg =
+            Message.newBuilder()
+                .setId(admittedEvent.getRequest().getMeta().getUpdateId() + "/request")
+                .setProtocolInstanceId(admittedEvent.getRequest().getMeta().getUpdateId())
+                .setEventId(event.getEventId())
+                .setBody(Any.pack(admittedEvent.getRequest()))
+                .build();
+        if (replaying) {
+          messages =
+              messages.stream()
+                  .map(
+                      m -> {
+                        if (m.getProtocolInstanceId().equals(msg.getProtocolInstanceId())) {
+                          // Check if we need to replace the update message synthesize from an
+                          // accepted event with the update message synthesize from an admitted
+                          // event
+                          return m.toBuilder().setBody(msg.getBody()).build();
+                        }
+                        return m;
+                      })
+                  .collect(Collectors.toList());
```

Comment:
Technically this is quadratic in the number of events -- does this worry you at all from a performance point of view? For example, suppose someone did a batch reset of many workflows that were accepting a high volume of updates (some sort of entity workflow pattern). That could send a post-reset spike of `UpdateAdmitted` events. Could a scenario like that result in a spike in worker CPU activity worth worrying about?

Reply:
Why is it quadratic and why over the number of events? A map is a linear operation and a collect is as well? very similar to your PR in core https://github.com/temporalio/sdk-core/pull/700/files#diff-fcb2b6f2d5782438cd1f138a0d39cc349ddfaf7e3f4049ad8a7499631cacc3f1R729

Reply:
This map/collect over messages is inside a for loop over events starting at [WorkflowStateMachines.java#L310](https://github.com/temporalio/sdk-java/blob/master/temporal-sdk/src/main/java/io/temporal/internal/statemachines/WorkflowStateMachines.java#L310), and the number of messages grows with the number of events, since we synthesize them from events. (I haven't looked at core, I believe you that my question could apply there too!)

Reply:
Oh the number of Update Admitted Events yes that makes sense, sorry I thought you meant over the total number of history events my bad. Yeah I could switch to storing messages in a map, hmm

Reply:
OK going to go back to the previous approach , not sure how big a problem issue this would really be, but neither solution is clearly better then the other.

Reply:
> very similar to your PR in core https://github.com/temporalio/sdk-core/pull/700/files#diff-fcb2b6f2d5782438cd1f138a0d39cc349ddfaf7e3f4049ad8a7499631cacc3f1R729

Right, my PR in core is using a hash map, so that seems reasonable to me if you did want to change it.